jl4-core/src/L4/TypeCheck.hs

```diff
@@ -936,74 +1003,35 @@ maybeGivethType :: Maybe (GivethSig Resolved) -> Check (Type' Resolved)
 maybeGivethType Nothing                  = fresh (NormalName "r") -- we have no obvious prefix?
 maybeGivethType (Just (MkGivethSig _ t)) = pure t -- type is already resolved
 
--- | Ensure that changes to the environment remain local to the passed computation.
--- Also scopes the error context.
---
-scope :: Check a -> Check a
-scope m = do
-  savedCtx <- use #errorContext
-  savedEnv <- use #environment
-  savedEi  <- use #entityInfo -- possibly not necessary, but also not harmful
-  a <- m
-  assign #errorContext savedCtx
-  assign #environment savedEnv
-  assign #entityInfo savedEi
-  pure a
-
-setErrorContext :: (CheckErrorContext -> CheckErrorContext) -> Check ()
-setErrorContext f =
-  modifying #errorContext f
-
 ensureDistinct :: NonDistinctContext -> [Name] -> Check ()
 ensureDistinct ndc ns = do
   let
     dups = filter ((>= 2) . length) (groupOn rawName (sortOn rawName ns))
   unless (null dups) $
     addError (NonDistinctError ndc dups)
 
--- | Makes the given named item known in the current scope,
--- with the given specification.
---
-makeKnown :: Resolved -> CheckEntity -> Check ()
-makeKnown a ce = do
-  -- traceM $ "Trying to make known " <> Text.unpack (prettyLayout a) <> ": " <> debugCandidate (u, n, ce)
-  modifying' #environment (Map.alter proc (rawName n))
-  modifying' #entityInfo  (Map.insert u (n, ce))
-  where
-    u :: Unique
-    n :: Name
-    (u, n) = getUniqueName a
-
-    proc :: Maybe [Unique] -> Maybe [Unique]
-    proc Nothing   = Just [u]
-    proc (Just xs) = Just (u : xs)
-
-makeKnownMany :: [Resolved] -> CheckEntity -> Check ()
-makeKnownMany rs ce =
-  traverse_ (flip makeKnown ce) rs
-
 checkExpr :: ExpectationContext -> Expr Name -> Type' Resolved -> Check (Expr Resolved)
-checkExpr ec (IfThenElse ann e1 e2 e3) t = softprune $ scope $ do
+checkExpr ec (IfThenElse ann e1 e2 e3) t = softprune $ do
   re <- checkIfThenElse ec ann e1 e2 e3 t
   let re' = setAnnResolvedType t re
   pure re'
-checkExpr ec (Consider ann e branches) t = softprune $ scope $ do
+checkExpr ec (Consider ann e branches) t = softprune $ do
```

Comment:
why did this `scope` disappear without replacement?

Reply:
I guess this was a bug before?

Reply:
All scopes are explicit now, there is no one `scope` function any more. `scope` has been replaced by `errorContext` and `extendKnown*`, which modify the `ErrorContext` and `Environment`/`EntityInfo` for the given scope.

Reply:
Yeah I mean that I understand but I still wonder what the scope that was there before was supposed to scope over…

Reply:
It was scoping the definitions of `Where` and `Lambda`, which both brought new binders into scope. This is now managed explicitly.